### PR TITLE
Tilde expansion

### PR DESCRIPTION
**Pull request by zackradisic**
### What does this PR do?

This adds basic support for [tilde expansion](https://pubs.opengroup.org/onlinepubs/009604499/utilities/xcu_chap02.html#tag_02_06_01) in Bun shell. 

"Basic" means that tilde is expanded in two scenarios:

```bash
# lone tilde expands to $HOME
echo ~

# tilde + path expands to $HOME/$PATH
echo ~/Documents
```

Also, this changes how JS strings/objects are substituted into the shell script. It now uses a non-printable char as a special token to denote a substituted JS string/object (it was previously using a tilde).

- [x] Documentation or TypeScript types (it's okay to leave the rest blank in this case)
- [x] Code changes

### How did you verify your code works?

- [x] I checked the lifetime of memory allocated to verify it's (1) freed and (2) only freed when it should be
- [x] I included a test for the new code, or an existing test covers it
- [x] JSValue used outside outside of the stack is either wrapped in a JSC.Strong or is JSValueProtect'ed
- [x] I wrote TypeScript/JavaScript tests and they pass locally (`bun-debug test test-file-name.test`)

### CODE REVIEW

File: src/shell/shell.zig [hunk 1]
```diff
@@ -2292,7 +2309,8 @@ pub fn NewLexer(comptime encoding: StringEncoding) type {
                 const escaped = input.escaped;
 
                 // Special token to denote substituted JS variables
-                if (char == '~') {
+                // we use 8 or \b which is a non printable char
+                if (char == 8) {
```
Comment (Jarred-Sumner) [inline src/shell/shell.zig]:
```suggestion
                if (char == SPECIAL_JS_CHAR) {
```

File: src/shell/shell.zig [hunk 2]
```diff
@@ -2152,8 +2169,8 @@ pub const LexError = struct {
     /// Allocated with lexer arena
     msg: []const u8,
 };
-pub const LEX_JS_OBJREF_PREFIX = "~__bun_";
-pub const LEX_JS_STRING_PREFIX = "~__bunstr_";
+pub const LEX_JS_OBJREF_PREFIX = &[_]u8{8} ++ "__bun_";
```
Comment (Jarred-Sumner) [inline src/shell/shell.zig]:
```suggestion
const SPECIAL_JS_CHAR = 8;
pub const LEX_JS_OBJREF_PREFIX = &[_]u8{SPECIAL_JS_CHAR} ++ "__bun_";
```

File: src/shell/shell.zig [hunk 3]
```diff
@@ -2152,8 +2169,8 @@ pub const LexError = struct {
...
+pub const LEX_JS_STRING_PREFIX = &[_]u8{8} ++ "__bunstr_";
```
Comment (Jarred-Sumner) [inline src/shell/shell.zig]:
```suggestion
pub const LEX_JS_STRING_PREFIX = &[_]u8{SPECIAL_JS_CHAR} ++ "__bunstr_";
```

**Comment by Jarred-Sumner** [comment]
instead of `8` we could use a named constant

otherwise LGTM

File: src/shell/interpreter.zig [hunk 4]
```diff
@@ -885,18 +885,11 @@ pub const Interpreter = struct {
...
-                    env.ref();
-                    return env;
-                }
-            }
-            return EnvStr.initSlice("unknown");
+            const env_var: ?EnvStr = brk: {
+                const static_str = if (comptime bun.Environment.isWindows) EnvStr.initSlice("USERPROFILE") else EnvStr.initSlice("HOME");
+                break :brk self.shell_env.get(static_str) orelse self.export_env.get(static_str);
+            };
+            return env_var orelse EnvStr.initSlice("unknown");
```
Comment (gvilums) [inline src/shell/interpreter.zig]: does it make sense to return the string `unknown`? If we expand `~` to `unknown` it's just going to lead to really confusing errors

**Review by gvilums — COMMENTED** [review]
Looks good except for one detail

**Comment by Jarred-Sumner** [comment]
`root` is usually `/root` on Linux

Do we need to do anything for that to work?

**Pull request merged**